### ISSUE TITLE
Issue with streaming traffic with MPTCP upstream implementation

### ISSUE THREAD

**Issue by abhinaba-fbr** [multipath-tcp/mptcp_net-next#326]
Hello,

I have been trying to stream MPEG-DASH traffic over MPTCP. Used a simple peer to peer topology with two pair of interfaces connected between the cilent and server. However, we have observed the following scenario -
	
1. Master subflow is getting established. Client requests for MPEG-DASH chunks and is being replied by the server over the master sublow.
2. We are able to add a new subflow sucessfully, however, the data is not being streamed over it.
3. The client sends GET get requests over the additional subflow but the server instead of replying with the DASH chunks FIN the connection.
4. We also tried to check the resilence by turning off the maters subflow's interface cards in the middle of streaming. Instead of switching to the additional subflow and continuing the stream, it totally stops playback.

Has anyone observed the same issue with DASH traffic in upstream implementation?
Is it an issue of MPTCP upstream implementation or does it require additional configuration from the server side in our experiment?

Thanks

**Comment by matttbe** [comment]
Hi @abhinaba-fbr,

> Used a simple peer to peer topology with two pair of interfaces connected between the cilent and server.

Which kernel versions are you using on both sides? The MPTCP dev is quite [active](https://github.com/multipath-tcp/mptcp_net-next/wiki#changelog), it is important to use a recent kernel, e.g. 6.x.
How did you configure MPTCP? endpoints? routing rules?

> We are able to add a new subflow sucessfully, however, the data is not being streamed over it.

Do you have a packet trace (tcpdump/wireshark) showing what is going on? `ip mptcp monitor` on both sides can also be helpful.
Do you have an established subflow but no data being exchanged? Did you flag endpoints as backup?

> The client sends GET get requests over the additional subflow but the server instead of replying with the DASH chunks FIN the connection.

So you do have data going over the additional subflow?
Does it close with a fastclose? Did you enabled the MPTCP checksum support?

> We also tried to check the resilence by turning off the maters subflow's interface cards in the middle of streaming. Instead of switching to the additional subflow and continuing the stream, it totally stops playback.

The previous steps seem to show there is something wrong somewhere so it is normal this doesn't work (configuration?)

> Has anyone observed the same issue with DASH traffic in upstream implementation?

I didn't hear anything but the application should not matter.
Did you modify the application to create an MPTCP socket or did you use mptcpize?

> Is it an issue of MPTCP upstream implementation or does it require additional configuration from the server side in our experiment?

Did you look at the checklist from our [wiki](https://github.com/multipath-tcp/mptcp_net-next/wiki#how-to-use-mptcp)?

**Comment by abhinaba-fbr** [comment]
Hello @matttbe, thanks for your reply.

> Which kernel versions are you using on both sides? The MPTCP dev is quite [active](https://github.com/multipath-tcp/mptcp_net-next/wiki#changelog), it is important to use a recent kernel, e.g. 6.x.
How did you configure MPTCP? endpoints? routing rules?

Initially we were using kernel v5.19. However after suggestion we upgraded the kernel version v6.0.9, ubuntu 22.10. Still we faced the similar issue. We configured endpoints using ip mptcp and also set the routing rules as per the [link](https://multipath-tcp.org/pmwiki.php/Users/ConfigureRouting) from multipath-tcp.org. 

> Do you have an established subflow but no data being exchanged? Did you flag endpoints as backup?

Yes additional subflow is being established, however it is being closed with MPTCP RST. We did not flag the endpoints as backup.

> Does it close with a fastclose? Did you enabled the MPTCP checksum support?

No, it does not close with fastclose, seems the subflows are being terminated by the server with MPTCP RST (error code - MPTCP-specific-error).  We also kept the checksum disabled.

We used the following topology with Server running on port 8000 -

![topo1 png](https://user-images.githubusercontent.com/53138315/208352572-68f30f7c-786c-4bfa-822e-20e8fb2052fd.png)

Mptcpize is used for creating the mptcp sockets for the specific applications. Here is the link to the pcap file we obtained.
1. Capturing at eth1a ([pcap](https://drive.google.com/drive/folders/1j4EXlNUudn8UpY5Lr5b1fQJc0rdmY1v-))
2. Capturing at eth1b ([pcap](https://drive.google.com/drive/folders/1j4EXlNUudn8UpY5Lr5b1fQJc0rdmY1v-))

I would specifically like you to check two subflows with port numbers (tcp.port==56411 and tcp.port==56769) on the AdditionalSubflow. On port 56411 the FIN is sent before the GET request is sent and then the connection is reset. At port 56769 the join connection fails due to MPTCP specific error.

@matttbe can you look into the pcaps and suggest what is going wrong? 

Thanks

**Comment by matttbe** [comment]
> > Which kernel versions are you using on both sides? The MPTCP dev is quite [active](https://github.com/multipath-tcp/mptcp_net-next/wiki#changelog), it is important to use a recent kernel, e.g. 6.x.
> > How did you configure MPTCP? endpoints? routing rules?
> 
> Initially we were using kernel v5.19. However after suggestion we upgraded the kernel version v6.0.9, ubuntu 22.10. Still we faced the similar issue. We configured endpoints using ip mptcp and also set the routing rules as per the [link](https://multipath-tcp.org/pmwiki.php/Users/ConfigureRouting) from multipath-tcp.org.

Thank you for having checked!

> > Do you have an established subflow but no data being exchanged? Did you flag endpoints as backup?
> 
> Yes additional subflow is being established, however it is being closed with MPTCP RST. We did not flag the endpoints as backup.

OK

> > Does it close with a fastclose? Did you enabled the MPTCP checksum support?
> 
> No, it does not close with fastclose, seems the subflows are being terminated by the server with MPTCP RST (error code - MPTCP-specific-error). We also kept the checksum disabled.

OK

> We used the following topology with Server running on port 8000 -
> 
> ![topo1 png](https://user-images.githubusercontent.com/53138315/208352572-68f30f7c-786c-4bfa-822e-20e8fb2052fd.png)

I guess you took the capture on the client side, right?
There are multiple paths that can lead to an `MPTCP-specific error (0x1)` but most of them are linked to "malformed" MPTCP header. It could be nice to instrument the kernel to find which one.
But with the view from the client, we cannot know where is the issue. For technical reasons, the kernel needs to send the SYN+ACK then the error. Do you have anything in your setup on the "second" interface that can mangle packets?

Maybe `nstat` before and after (on the server side) could help.
Could you also enable MPTCP Checksum? Maybe the issue will be detected earlier?

> Mptcpize is used for creating the mptcp sockets for the specific applications. Here is the link to the pcap file we obtained.
> 
>     1. Capturing at eth1a ([pcap](https://drive.google.com/drive/folders/1j4EXlNUudn8UpY5Lr5b1fQJc0rdmY1v-))
> 
>     2. Capturing at eth1b ([pcap](https://drive.google.com/drive/folders/1j4EXlNUudn8UpY5Lr5b1fQJc0rdmY1v-))
> 
> 
> I would specifically like you to check two subflows with port numbers (tcp.port==56411 and tcp.port==56769) on the AdditionalSubflow. On port 56411 the FIN is sent before the GET request is sent and then the connection is reset. At port 56769 the join connection fails due to MPTCP specific error.

For the MPTCP specific error, it is interesting. As said above, we need to find exactly why the server didn't like the connection request.
It could also be due to the fact the kernel receive the request but didn't find any associated MPTCP connection (different namespace? etc.)

For the other one, it is strange to receive data after a FIN has been sent and acked.

**Comment by abhinaba-fbr** [comment]
> I guess you took the capture on the client side, right?

Yes

>  Do you have anything in your setup on the "second" interface that can mangle packets?

Well we are actually using network namespaces to emulate the experiments and used netem to modify the device bitrate and delay. In the server side we did the following endpoint configuration -

```bash
$ mptcp endpoint flush
$ mptcp limits set subflow 3 add_addr_accepted 3
$ mptcp endpoint add 192.168.0.2 dev eth2b id 1 signal
```
And on the client side we did the following endpoint configuration -
```bash
$ mptcp endpoint flush
$ mptcp limits set subflow 3 add_addr_accepted 3
$ mptcp endpoint add 192.168.0.1 dev eth1b id 2 fullmesh
```
> Maybe nstat before and after (on the server side) could help.

I did as you mentioned. Here is the file of the nstat before and after the stream.
Before -
```bash
#kernel
```
After -
```bash
#kernel
IpInReceives                    72367              0.0
IpInDelivers                    72367              0.0
IpOutRequests                   72641              0.0
TcpActiveOpens                  20                 0.0
TcpPassiveOpens                 383                0.0
TcpAttemptFails                 196                0.0
TcpEstabResets                  159                0.0
TcpInSegs                       72367              0.0
TcpOutSegs                      84112              0.0
TcpRetransSegs                  15                 0.0
TcpOutRsts                      407                0.0
Ip6InReceives                   20                 0.0
Ip6InNoRoutes                   1                  0.0
Ip6OutRequests                  23                 0.0
Ip6InMcastPkts                  19                 0.0
Ip6OutMcastPkts                 23                 0.0
Ip6InOctets                     1272               0.0
Ip6OutOctets                    1500               0.0
Ip6InMcastOctets                1196               0.0
Ip6OutMcastOctets               1500               0.0
Ip6InNoECTPkts                  20                 0.0
Icmp6OutMsgs                    23                 0.0
Icmp6OutRouterSolicits          12                 0.0
Icmp6OutNeighborSolicits        2                  0.0
Icmp6OutMLDv2Reports            9                  0.0
Icmp6OutType133                 12                 0.0
Icmp6OutType135                 2                  0.0
Icmp6OutType143                 9                  0.0
TcpExtEmbryonicRsts             85                 0.0
TcpExtPruneCalled               37377              0.0
TcpExtTW                        26                 0.0
TcpExtDelayedACKs               53745              0.0
TcpExtDelayedACKLocked          1                  0.0
TcpExtTCPPureAcks               13144              0.0
TcpExtTCPSackRecovery           8                  0.0
TcpExtTCPSACKReorder            2                  0.0
TcpExtTCPTSReorder              8                  0.0
TcpExtTCPPartialUndo            8                  0.0
TcpExtTCPFastRetrans            15                 0.0
TcpExtTCPDSACKRecv              13                 0.0
TcpExtTCPAbortOnData            11                 0.0
TcpExtTCPAbortOnClose           159                0.0
TcpExtTCPDSACKIgnoredNoUndo     1                  0.0
TcpExtTCPSackMerged             16                 0.0
TcpExtTCPSackShiftFallback      9                  0.0
TcpExtTCPOrigDataSent           24355              0.0
TcpExtTCPHystartDelayDetect     115                0.0
TcpExtTCPHystartDelayCwnd       5733               0.0
TcpExtTCPDelivered              24366              0.0
TcpExtTCPDSACKRecvSegs          15                 0.0
IpExtInOctets                   15360513           0.0
IpExtOutOctets                  37741128           0.0
IpExtInNoECTPkts                72367              0.0
MPTcpExtMPCapableSYNRX          234                0.0
MPTcpExtMPCapableACKRX          234                0.0
MPTcpExtMPTCPRetrans            4                  0.0
MPTcpExtMPJoinSynRx             234                0.0
MPTcpExtMPJoinAckRx             58                 0.0
MPTcpExtMPJoinAckHMacFailure    91                 0.0
MPTcpExtEchoAdd                 234                0.0
MPTcpExtMPRstTx                 159                0.0
MPTcpExtRcvWndShared            58003              0.0

```
From here it seems, the MPTCP specific error is due to HMAC failure during join connection. 
Surprising when I removed configuring the devices with netem, no more MPTCP specific error was observed. Can netem interfere with the implementation to mangle the header?

However, the other issue (FIN before sending any HTTP data) is still continuing to happen, due to which the subflow is not begin used.

Has it something to do with MPTCP schedulers? Does current implementation has schedulers which can distribute the packets evenly on the interface or something like redundent scheduler as mentioned in RFC?

**Comment by matttbe** [comment]
> And on the client side we did the following endpoint configuration -
> 
> ```shell
> $ mptcp endpoint flush
> $ mptcp limits set subflow 3 add_addr_accepted 3
> $ mptcp endpoint add 192.168.0.1 dev eth1b id 2 fullmesh
> ```

Mmh, I don't know if we properly handle having only the `fullmesh` flag without the `subflow` one. The original design was to do:

```
mptcp endpoint add 192.168.0.1 dev eth1b subflow fullmesh
```

(you don't need to give the id)

But from what I saw on the traces (if I remember well), it is always the server side which is abruptly closing connections

> Here is the file of the nstat before and after the stream.

Thank you for sharing this. By chance, can you easily reproduce the issue with a single connection?
Because there are quite a lot going on there when looking at nstat. Indeed there are the HMAC failure but also: `TcpAttemptFails`, `TcpEstabResets`, `TcpExtEmbryonicRsts`, `TcpExtTCPAbortOnData`, `TcpExtTCPAbortOnClose` (reset/aborts/... maybe normal and requested by the application?), `TcpExtPruneCalled` (not enough memory or undersized buffers?)

> Surprising when I removed configuring the devices with netem, no more MPTCP specific error was observed. Can netem interfere with the implementation to mangle the header?

That's surprising indeed. What was your TC config? I guess netem would drop packets, not a part of them. You can mangle packets with TC but I guess you didn't do that.

> However, the other issue (FIN before sending any HTTP data) is still continuing to happen, due to which the subflow is not begin used.

When I looked at the traces, I saw a lot of exchange data, then the HTTP GET, then the FIN. It looks a bit strange but maybe normal?
Did you check what was the corresponding traffic on the other subflow? Does it also behave like that?

Because there are a lot of reset, I wonder if you don't have the same issue as reported in #295, solved in Linux v6.1.

What could help is to manage to reproduce it with the minimum: one connection, nstat and traces on both sides, maybe with another tool.

> Has it something to do with MPTCP schedulers?

The packet schedulers will just distributes packets, not close paths. This is done by the Path Manager you configured with `ip mptcp`

> Does current implementation has schedulers which can distribute the packets evenly on the interface or something like redundent scheduler as mentioned in RFC?

The work is ongoing to support these very specific schedulers. You probably don't need them, they are only useful for very specific use-cases: round-robin is just for the experimentations why the redundant scheduler is to keep a low latency if only a few data are sent (not sure TCP should be used if latency is so important because there will be retransmissions anyway if packets are lost)

**Comment by abhinaba-fbr** [comment]
Hello @matttbe 
Thanks for your reply.

> Mmh, I don't know if we properly handle having only the fullmesh flag without the subflow one. The original design was to do:

```
mptcp endpoint add 192.168.0.1 dev eth1b subflow fullmesh
```

Okay, we changed using only `fullmesh` to `subflow fullmesh`.

> That's surprising indeed. What was your TC config?

We used the following `tc` configuration -

```bash
tc qdisc add dev eth1b root netem delay 10ms rate 5mbit
```
This should not create an HMAC failure, since it was working fine as well when I was using MPTCP for iperf3 traffic. In this https://github.com/multipath-tcp/mptcp_net-next/issues/331#issuecomment-1361552212 @pabeni mentioned a similar scenario. Are these two issues related? 

> Thank you for sharing this. By chance, can you easily reproduce the issue with a single connection?

We can do so, but we found the issue was actually on the server side as you mentioned. 

> But from what I saw on the traces (if I remember well), it is always the server side which is abruptly closing connections

We were using a server which was only supporting HTTP/1.0, due to which a new connection was being used for each request and data reply was not being sent on the subflow. On updating the server to support HTTP/1.1 the issue got resolved.

Now both the links are being used as expected. 

Thanks for providing us with some useful pointers on debugging the issue. 

However, on changing the server `tc` is also not causing any trouble now.
I m unable to understand the reason behind HMAC failure (MPTCP-specific-error)?

**Comment by matttbe** [comment]
> This should not create an HMAC failure, since it was working fine as well when I was using MPTCP for iperf3 traffic. In this [#331 (comment)](https://github.com/multipath-tcp/mptcp_net-next/issues/331#issuecomment-1361552212) @pabeni mentioned a similar scenario. Are these two issues related?

Indeed, it should not create HMAC failure. I will check with Paolo later on what was exactly causing issues. I suspect it is because when using `nc`, the server side closes the "listen socket" after the `accept()`: by doing that, the server side cannot accept new subflows and probably causing an HMAC failure.

> We were using a server which was only supporting HTTP/1.0, due to which a new connection was being used for each request and data reply was not being sent on the subflow. On updating the server to support HTTP/1.1 the issue got resolved.

That's good it is fixed but it is not clear to me why the second subflow was not used before. Is it because the connections were too short? Or because the client was acting differently not allowing new subflows? (e.g. closing the listen socket?)

> However, on changing the server tc is also not causing any trouble now.

Strange indeed. I'm not sure what to say: the ideal would be to reproduce the issue in a "small" setup and/or debug the kernel to understand what's going on.

**Comment by abhinaba-fbr** [comment]
> That's good it is fixed but it is not clear to me why the second subflow was not used before. Is it because the connections were too short? Or because the client was acting differently not allowing new subflows? (e.g. closing the listen socket?)

Yes, with HTTP/1.0 the connections were indeed short lived, we increased the ```stale_loss_cnt``` variable such that the subflow is not declared stale. And also we tried to send the data as early as possible in the subflow by trurning off the master subflow interface cards. According to me, since HTTP/1.0 is not a persistent connection, as soon as we turn the interface cards, master subflow does not get responded with the desired packets (HTTP ok) and it closes the streaming. However the MPTCP sockets are still alive. HTTP/1.1 overcomes this by its persistent connection features, the switching over takes smoothly, even if for sometime no desired response is received.

> Strange indeed. I'm not sure what to say: the ideal would be to reproduce the issue in a "small" setup and/or debug the kernel to understand what's going on.

Yes that is strange, we will try to reproduce the issue with a small setup (single connection), and get back to you.

Thanks

**Comment by matttbe** [comment]
Hi @abhinaba-fbr : did you manage to reproduce the issue with a small and controlled setup?

I'm not sure what we should do with this ticket.

**Comment by abhinaba-fbr** [comment]
Hello @matttbe 

Yes, I tried to reconstruct the issue. However, I could not find any MPTCP specific error. Except that in most cases I observe that application data (video chunk) is not scheduled at the second flow (MP_JOIN). This can be due to the fact that MPTCP default scheduler only schedules packet over lowest RTT link. 

Apart from that, what I observed that the primary issue was with the HTTP server itself. Since, HTTP 1.0 creates multiple sockets. Once the primary link (MP_CAPABLE) goes down. It fails to create the MP_CAPABLE connection itself. Hence streaming fails. However, it was fixed with HTTP/1.1 as it reuses the socket.

HMAC failure issue was the only MPTCP specific error is observed when we configure the links with tc and netem. However, I haven't tried to debug the kernel to see, what goes wrong on using netem.

I think we can close this ticket as the primary issue with streaming is resolved. Regarding the HMAC issue, maybe we can create a separate ticket for that.

**Comment by matttbe** [comment]
Thank you for the update!

Yes, it will be clearer to open separated issues because there were a few different discussions going on and that's confusing for the person who wants to look at a fix.

(I'm still trying to find the link with the HMAC issue and netem, probably some MIB counters are increased in wrong cases)
